backend/service/k8s/statefulset.go

```diff
@@ -0,0 +1,123 @@
+package k8s
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/client-go/util/retry"
+
+	k8sapiv1 "github.com/lyft/clutch/backend/api/k8s/v1"
+)
+
+func (s *svc) DescribeStatefulSet(ctx context.Context, clientset, cluster, namespace, name string) (*k8sapiv1.StatefulSet, error) {
+	cs, err := s.manager.GetK8sClientset(clientset, cluster, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	statefulSets, err := cs.AppsV1().StatefulSets(cs.Namespace()).List(ctx, metav1.ListOptions{
+		FieldSelector: "metadata.name=" + name,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(statefulSets.Items) == 1 {
+		return ProtoForStatefulSet(cs.Cluster(), &statefulSets.Items[0]), nil
+	} else if len(statefulSets.Items) > 1 {
+		return nil, fmt.Errorf("Located multiple StatefulSets")
+	}
+
+	return nil, fmt.Errorf("Unable to locate StatefulSet")
+}
+
+func ProtoForStatefulSet(cluster string, statefulSet *appsv1.StatefulSet) *k8sapiv1.StatefulSet {
+	clusterName := statefulSet.ClusterName
+	if clusterName == "" {
+		clusterName = cluster
+	}
+	return &k8sapiv1.StatefulSet{
+		Cluster:     clusterName,
+		Namespace:   statefulSet.Namespace,
+		Name:        statefulSet.Name,
+		Labels:      statefulSet.Labels,
+		Annotations: statefulSet.Annotations,
+	}
+}
+
+func (s *svc) UpdateStatefulSet(ctx context.Context, clientset, cluster, namespace, name string, fields *k8sapiv1.UpdateStatefulSetRequest_Fields) error {
+	cs, err := s.manager.GetK8sClientset(clientset, cluster, namespace)
+	if err != nil {
+		return err
+	}
+
+	getOpts := metav1.GetOptions{}
+	oldStatefulSet, err := cs.AppsV1().StatefulSets(cs.Namespace()).Get(ctx, name, getOpts)
+	if err != nil {
+		return err
+	}
+
+	newStatefulSet := oldStatefulSet.DeepCopy()
+	mergeStatefulSetLabelsAndAnnotations(newStatefulSet, fields)
+
+	patchBytes, err := generateStatefulSetStrategicPatch(oldStatefulSet, newStatefulSet)
+	if err != nil {
+		return err
+	}
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		_, err := cs.AppsV1().StatefulSets(cs.Namespace()).Patch(ctx, oldStatefulSet.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+		return err
+	})
+	return retryErr
+}
+
+func mergeStatefulSetLabelsAndAnnotations(statefulSet *appsv1.StatefulSet, fields *k8sapiv1.UpdateStatefulSetRequest_Fields) {
```

Comment:
To reduce code duplication lets find a more generic way to merge labels and annotations.

Reply:
re-worked this to use [labels.Merge](https://github.com/kubernetes/apimachinery/blob/master/pkg/labels/labels.go#L114) for a more clean way of doing it with lesser code duplication

Reply:
if this was reworked i'm not seeing the new code

Reply:
hmmm...looks like the rebase accidentally removed the changes :(

Reply:
@danielhochman rework can be viewed [here](https://github.com/lyft/clutch/pull/593/commits/ef6abbedb304dcf2e063eaae5eb1784a629d3e2a). Used `labels.Merge` to do this in a more generic way. Since we have to update the labels in 2 locations in the object thought it would be cleaner to let each resource have its own method with lesser code duplicated between the methods.